File: go/bank-account/bank_account.go
package account

import "sync"

type Account struct {
	mu    sync.Mutex
	money int64
	open bool
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	return &Account{money: amount, open: true}
}

func (a *Account) Balance() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.open {
		return a.money, true
	} else {
		return 0, false
	}
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.open || a.money + amount < 0 {
		return 0, false
	} else {
		a.money += amount
		return a.money, true
	}
}

func (a *Account) Close() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.open {
		a.open = false
		money := a.money
		a.money = 0
		return money, true
	} else {
		return 0, false
	}
}

File: javascript/bank-account/bank-account.js
export class BankAccount {
  constructor() {
    [this._open, this._balance] = [false, 0];
  }

  open() {
    if (this._open) {
      throw new ValueError();
    }
    this._open= true;
  }

  close() {
    if (!this._open) {
      throw new ValueError();
    }
    [this._open, this._balance] = [false, 0];
  }

  deposit(amount) {
    if (!this._open || amount < 0) {
      throw new ValueError();
    }
    this._balance += amount;
  }

  withdraw(amount) {
    if (!this._open || amount < 0 || amount > this._balance) {
      throw new ValueError();
    }
    this._balance -= amount;
  }

  get balance() {
    if (!this._open) {
      throw new ValueError();
    }
    return this._balance;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
